src/code.cloudfoundry.org/loggregator/plumbing/envelope_averager.go

```diff
@@ -0,0 +1,81 @@
+package plumbing
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+// EnvelopeAverager implements google.golang.org/grpc/stats.Handler and keeps
+// track of the average envelope size emitted by a grpc client. It should
+// be constructed with NewEnvelopeAverager.
+//
+// Note: Track is on a very "hot" path.
+// Care should be taken while altering any algorithms, and performance
+// must be considered.
+type EnvelopeAverager struct {
+	// storage must be accessed via atomics
+	storage countAndTotal
+}
+
+// NewEnvelopeAverager creates a new EnvelopeAverager.
+func NewEnvelopeAverager() *EnvelopeAverager {
+	return &EnvelopeAverager{}
+}
+
+// Track takes the given envelope size (in bytes) to use in the current
+// average calculation. It can be called by several go-routines.
+func (a *EnvelopeAverager) Track(envelopeSize int) {
+	for {
+		current := atomic.LoadUint64((*uint64)(&a.storage))
+		newValue := countAndTotal(current).encodeDelta(1, uint32(envelopeSize))
+
+		if atomic.CompareAndSwapUint64((*uint64)(&a.storage), current, uint64(newValue)) {
+			return
+		}
+	}
+}
+
+// Start invokes the given callback at the set interval.
```

Comment:
This could be improved a bit `Start invokes the given callback with the average envelope size of the past interval.`